### DISCUSSION TITLE
Using idom from FastAPI

### DISCUSSION THREAD

**Discussion by seeclearvision** [idom-team/idom#650] [Question]
I have a couple of questions regarding the use of idom from FastAPI?

It seems to be giving me the following error:

RuntimeError: asyncio.run() cannot be called from a running event loop
/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/uvicorn/protocols/http/h11_impl.py:384: RuntimeWarning: coroutine 'Server.serve' was never awaited

I don't know how to get it to work with FastAPI.

Any help is greatly appreciated.

Cheers,
Dev.

**Comment by rmorshea** [comment]
Do you think you could post the code sample that's producing this error? Also, have you checked out the documentation on [running IDOM](https://idom-docs.herokuapp.com/docs/getting-started/running-idom.html)? There's a number of code samples there. If any of them fail you should post back here with the code and associated tracebacks you're seeing.

**Reply by seeclearvision** [reply]
import idom
from idom import component, html

import uvicorn
from fastapi import FastAPI

@idom.component
async def MyHeader():
    return idom.html.h1("This is my header!")

app = FastAPI()

@app.get("/")
async def root():
    return {"message": "Hello World"}

@app.get("/myheader")
async def display_myheader():
    return idom.run(MyHeader)

if __name__ == "__main__":
    uvicorn.run(host="0.0.0.0", port=8080)

**Reply by Archmonger** [reply]
`idom.run()` is a development shortcut. It is used to start up the webserver for currently installed framework. In this case, it would auto launch the FastAPI development server. See [the docs](https://idom-docs.herokuapp.com/docs/getting-started/running-idom.html#running-idom) for how `idom.run()` is typically used.

Thus, calling `idom.run()` within the context of a FlaskAPI route does not make sense.

If you want to add IDOM to an existing project, and then launch that project using Hypercorn you will need to look the docs for [adding IDOM to an existing web server](https://idom-docs.herokuapp.com/docs/getting-started/running-idom.html#add-to-an-existing-web-server) or [embedding into an existing page](https://idom-docs.herokuapp.com/docs/getting-started/running-idom.html#embed-in-an-existing-webpage).

**Reply by seeclearvision** [reply]
Thanks @Archmonger.
Let me check out the link. Appreciate it guys.

**Reply by seeclearvision** [reply]
My apologies. The example uses 'Sanic'. How do we do it using uvicorn?

**Reply by Archmonger** [reply]
Sanic is a _**web framework**_, much like FastAPI.
Uvicorn is a _**webserver**_, which can run code from either Sanic or FastAPI.

I would recommend your first step to be follow the [running IDOM docs](https://idom-docs.herokuapp.com/docs/getting-started/running-idom.html#running-idom) to ensure FastAPI is working with IDOM.

Once you get things running, you can transition to using your own custom `FastAPI` application, rather than the one automatically generated by `idom.run()`. The [embed docs](https://idom-docs.herokuapp.com/docs/getting-started/running-idom.html#embed-in-an-existing-webpage) mention how to do it with `Sanic`, but the philosophy remains the same for FastAPI.

---

_Note: @rmorshea the docs need some touchup on running IDOM. See[ #654](https://github.com/idom-team/idom/discussions/654)_

**Reply by seeclearvision** [reply]
Thanks for the concise explanation Mark.

**Reply by seeclearvision** [reply]
Hi Mark/Ryan,
Would you happen to have a small app to kickstart an idom web app with a simple authentication?

**Reply by Archmonger** [reply]
Authentication will depend entirely on your _**web framework**_ and is not handled by IDOM. To oversimplify, the job of IDOM is simply to plop things on the web page. Everything else is handled by your web framework of choice.

For example, I personally use IDOM with Django, so I use the [Django authentication system](https://docs.djangoproject.com/en/4.0/topics/auth/).

I personally don't use FastAPI so I'm not entirely sure what the land of FastAPI user authentication looks like. I'm going to assume, based on the name _FastAPI_, it's probably not designed for convenient user authentication.

On the flip side, I'm confident user authentication would be [fairly standard in Flask](https://flask-login.readthedocs.io/en/latest/).

**Reply by rmorshea** [reply]
@seeclearvision just to follow up, have you managed to get things working?

**Reply by rmorshea** [reply]
I'd try taking a look at this example [here](https://idom-docs.herokuapp.com/docs/getting-started/running-idom.html#add-to-an-existing-web-server). The only difference for FastAPI is in the import of the server implementation and how you actually run the server.

```python
# main.py

from fastapi import FastAPI

from idom import component, html
from idom.server.fastapi import PerClientStateServer, Config

app = FastAPI()


@component
def IdomView():
    return html.h1("This is an IDOM App")


PerClientStateServer(IdomView, app=app, config=Config(url_prefix="app"))
```

Then in your terminal run:

```bash
uvicorn main:app
```

In this scenario, all IDOM is doing is attaching itself to the `app`. Then, it's up to you do actually run the `app`. For FastAPI, [you can do that with uvicorn](https://fastapi.tiangolo.com/tutorial/first-steps/). Another way to think about this is that IDOM is just adding another route to `app`.

**Reply by rmorshea** [reply]
If you're still having trouble, book a slot during office hours: https://github.com/idom-team/idom/discussions/312
